lib/luaxtable.c

```diff
@@ -32,48 +14,150 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret)
 {
-	return true;
+	int ret = -1;
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
```

Comment:
you can just use https://www.lua.org/manual/5.4/manual.html#lua_remove here, right?

Reply:
The stack contains : | base | arg1 |... | argn | table | function |
We need : | base | function |arg1 |... | argn |

We need to shift the function below args and remove the table from the stack. So, if we use remove, it would just remove the table, we still need to move the function below args.

Reply:
ah.. I see! ack